lib/explorer/query.ex

```diff
@@ -257,17 +257,120 @@ defmodule Explorer.Query do
+
+      # Callback equivalent
+      Explorer.DataFrame.mutate_with(df, fn query_frame ->
+        lazy_column_a = query_frame["a"]
+        lazy_column_b = Explorer.Series.add(lazy_column_a, 1)
+        [b: lazy_column_b]
+      end)
+
+  Although each verb expects a slightly different return value from their
+  callback, all expect a `%Explorer.Backend.QueryFrame{}` as the input. The
+  `%Explorer.Backend.QueryFrame{}` is a special struct which acts like an
+  `%Explorer.DataFrame{}` except that the `Access` behaviour is defined to
+  return _lazy_ versions of its columns.
+
+  To see this in action, we can work directly with an
+  `%Explorer.Backend.QueryFrame{}` since the `*_with` verbs are also designed to
+  accept the outputs of their callbacks:
+
+      # QueryFrame equivalent
+      df = Explorer.DataFrame.new(a: [1, 2, 3])
+      qf = Explorer.Query.new(df)
+      lazy_column_a = qf["a"]
+      lazy_column_b = Explorer.Series.add(lazy_column_a, 1)
+      Explorer.DataFrame.mutate_with(df, b: lazy_column_b)
+
+  You generally don't want to build a `%Explorer.Backend.QueryFrame{}` directly
+  like this because it's not convenient to work with. For one, the syntax is
+  verbose. For another, an individual `%Explorer.Backend.QueryFrame{}` can
+  become outdated:
+
+      # Callback chain
+      df = Explorer.DataFrame.new(a: [1, 2, 3])
+
+      df
+      |> Explorer.DataFrame.mutate(b: a + 1)
+      |> Explorer.DataFrame.mutate(c: b + 1)
+
+      # QueryFrame
+      df = Explorer.DataFrame.new(a: [1, 2, 3])
+      qf = Explorer.Query.new(df)
+
+      df
+      |> Explorer.DataFrame.mutate_with(b: Explorer.Series.add(qf["a"], 1))
+      # Work work! The original `qf` doesn't have a column `b`.
+      |> Explorer.DataFrame.mutate_with(c: Explorer.Series.add(qf["b"], 1))
+
+  When designing the `Explorer.Query` interface, we prioritized correctness.
+  E.g. we wanted to prevent users from attempting to access and/or manipulate
+  columns which didn't exist. So if you tried:
+
+      Explorer.DataFrame.new(a: [1, 2, 3])
+      |> Explorer.DataFrame.mutate_with(fn query_frame ->
+        not_a_column = query_frame["b"]
+        [c: Explorer.Series.add(not_a_column, 1)]
+      end)
+
+  You're immediately hit with an `ArgumentError`:
+
+      ** (ArgumentError) could not find column name "b". The available columns are: ["a"].
+      If you are attempting to interpolate a value, use ^b.
+
+  However, this meant that a new `%Explorer.Backend.QueryFrame{}` struct must be
+  built for every operation to ensure the columns and dtypes are valid. Hence
+  the callback interface: rather than forcing you to create a
+  `%Explorer.Backend.QueryFrame{}` every time, you write a callback instead that
+  accepts the fresh `%Explorer.Backend.QueryFrame{}` as input.
+
+  There are some limited instances where it's more convenient to work with
+  `Explorer.Backend.QueryFrame{}`s. If you want to re-use a lazy series, you can
```

Comment:
This section is about implementation details but you are giving advice to users on how to use part of the library. I recommend extracting this out and adding documentation to the `new` function, explaining when you should prefer `Explorer.Query.new` over the callback/macro API.